---
title: 'PIM inhibition on formation of PMAr-RS-TAMs (RS-PMAr 5 d): differential gene analysis'
author: "Michal Pawlak"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float: true
theme: cosmo

---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      fig.width = 8,
                      fig.asp = 0.618,
                      out.width = "90%",
                      fig.align = "center",
                      message=FALSE, 
                      warning=FALSE)

suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("ggrepel"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("DT"))
suppressPackageStartupMessages(library("sva"))
suppressPackageStartupMessages(library("Biobase"))
suppressPackageStartupMessages(library("DESeq2"))
suppressPackageStartupMessages(library("pheatmap"))
suppressPackageStartupMessages(library("RColorBrewer"))
suppressPackageStartupMessages(library("reshape2"))
suppressPackageStartupMessages(library("ggpubr"))
suppressPackageStartupMessages(library("org.Hs.eg.db"))
suppressPackageStartupMessages(library("tibble"))
suppressPackageStartupMessages(library("EnhancedVolcano"))
suppressPackageStartupMessages(library("magrittr"))
suppressPackageStartupMessages(library("msigdbr"))
suppressPackageStartupMessages(library("fgsea"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("purrr"))
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("UpSetR"))
suppressPackageStartupMessages(library("Vennerable"))

```

## Differential expression analysis

Differential gene expression analysis was performed with `DESeq2` with default parameters. Rows with padj > 0.05 were filtered out and the tables below represent significantly differential genes.

```{r deg}

dds <- readRDS(file = "../data/dds_mem_5d.RDS")
vsd <- readRDS(file = "../data/vsd_mem_5d.RDS")

write.csv(as.data.frame(assay(vsd)), file = "../data/tables/vsd_norm_counts_exp_2_PMA_L1236_MEN_3uM_d5.csv")

dds <- DESeq(dds)

#creating contrasts for DEG
contrasts <-
  list(
    DMSOvsMEM = c("treatment", "MEN", "DMSO"))

# colData(dds)$stim_2 <- factor(colData(dds)$stim_2)

DEG <- list()
for (i in 1:length(contrasts)) {
  DEG[[i]] <-
    as.data.frame(results(dds,
                          contrast = contrasts[[i]],
                          independentFiltering = TRUE, cooksCutoff = TRUE)) %>% tibble::rownames_to_column("SYMBOL") %>%
    dplyr::select(SYMBOL, log2FoldChange, pvalue, padj)
}

names(DEG) <- names(contrasts)

write.csv(DEG[[1]], file = "../data/tables/deg_exp_2_PMA_L1236_MEN_3uM_d5.csv")

saveRDS(DEG, "../data/DEG_MEN_3uM_d5.RDS")
```

### Volcano plots

Volcano plots are provided for each contrast with coloring as indicated on the legend. Cutoffs for coloring are: FC = 1, padj = 0.05. Number of up/downregulated genes is indicated for significantly changed genes (padj < 0.05).

```{r volcano}

res <- bind_rows(DEG, .id = "sample")

write_csv(res, path = "../data/tables/deg_exp_2_mem_d5.csv")

temp1 <- nrow(dplyr::filter(res, sample == "DMSOvsMEM", log2FoldChange<0, padj<0.05))
temp2 <- nrow(dplyr::filter(res, sample == "DMSOvsMEM", log2FoldChange>0, padj<0.05))


dat_text <- data.frame(label = c(paste(temp1, "\ndownregulated"), 
                                 paste(temp2, "\nupregulated")),
                       sample   = c("DMSOvsMEM", 
                                 "DMSOvsMEM"),
                    x = rep(c(-5, 5)),
                    y = rep(27, 2))

p1 <- EnhancedVolcano(res,
    lab = res$SYMBOL,
    selectLab = NA,
    x = 'log2FoldChange',
    y = 'padj',
    FCcutoff = 1,
    title = NULL,
    subtitle = NULL,
    pCutoff = 0.05,
    xlim = c(-7,7),
    ylim = c(0, 30),
    drawConnectors = TRUE,
    legendPosition = "bottom",
    legendLabSize = 10) +
    facet_wrap(~sample) + 
    theme(strip.background = element_rect(color="black", fill= "white"),
          text = element_text(size = 18)) + 
    geom_text(data = dat_text, mapping = aes(x = x, y = y, label = label))

p1
```


### Top differential genes heatmaps

To better visualize variable genes in each comparison, we are going to have a look at 25 most upregulated and 25 most downregulated genes in each contrasts and plot the valuse as a heatmap of sclaled (z-score) vsd values for each biological replicate.

```{r heatmaps_top, fig.asp=1}

topHeatmap <- function(x) {

top <- list()
for (i in 1:length(DEG)) {
top[[i]] <- rbind(top_n(DEG[[i]], x, log2FoldChange), top_n(DEG[[i]], -x, log2FoldChange))
}

vsd_table <- as.data.frame(assay(vsd))
vsd_scaled <- t(scale(t(vsd_table)))
colnames(vsd_scaled) <- vsd$treatment

pheatmap(
  vsd_scaled[rownames(vsd_scaled) %in% top[[1]]$SYMBOL, ],
  show_rownames = TRUE,
  cluster_cols = FALSE,
  cellwidth = 10,
  cluster_rows = TRUE,
  scale = "none",
  angle_col = 270,
  labels_row = top[[1]]$SYMBOL,
  main = paste0("top ", x, " up/down for DMSO vs MEM (vsd z-score)"))

}

topHeatmap(25)
```


```{r info}
sessionInfo()
```
