---
title: 'PIM inhibitor treatment of PMAr-RS-TAM for 24 hours: differential gene analysis'
author: "Michal Pawlak"
date: "`r Sys.Date()`"
output: 
  html_document:
    code_folding: hide
    toc: true
    toc_depth: 3
    toc_float: true
theme: cosmo

---

```{r setup, echo = FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      fig.width = 8,
                      fig.asp = 0.618,
                      out.width = "90%",
                      fig.align = "center",
                      message=FALSE, 
                      warning=FALSE)

suppressPackageStartupMessages(library("ggplot2"))
suppressPackageStartupMessages(library("ggrepel"))
suppressPackageStartupMessages(library("dplyr"))
suppressPackageStartupMessages(library("DT"))
suppressPackageStartupMessages(library("sva"))
suppressPackageStartupMessages(library("Biobase"))
suppressPackageStartupMessages(library("DESeq2"))
suppressPackageStartupMessages(library("pheatmap"))
suppressPackageStartupMessages(library("RColorBrewer"))
suppressPackageStartupMessages(library("reshape2"))
suppressPackageStartupMessages(library("ggpubr"))
#suppressPackageStartupMessages(library("org.Hs.eg.db"))
suppressPackageStartupMessages(library("tibble"))
suppressPackageStartupMessages(library("EnhancedVolcano"))
suppressPackageStartupMessages(library("magrittr"))
suppressPackageStartupMessages(library("msigdbr"))
#suppressPackageStartupMessages(library("fgsea"))
suppressPackageStartupMessages(library("stringr"))
suppressPackageStartupMessages(library("purrr"))
suppressPackageStartupMessages(library("readr"))
suppressPackageStartupMessages(library("UpSetR"))
suppressPackageStartupMessages(library("Vennerable"))

```

## Differential expression analysis

Differential gene expression analysis was performed with `DESeq2` with default parameters. Rows with padj > 0.05 were filtered out and the tables below represent significantly differential genes.

```{r deg}

dds <- readRDS(file = "../data/dds_MEM_24h.RDS")
vsd <- readRDS(file = "../data/vsd_MEM_24h.RDS")

write.csv(as.data.frame(assay(vsd)), file = "../data/tables/vsd_norm_counts_exp_2_MEM_24h.csv")

dds <- DESeq(dds)

#creating contrasts for DEG
contrasts <-
  list(
    DMSOvsMEN1703_3uM = c("group", "MEN1703_3uM", "DMSO"),
    DMSOvsMEN1703_5uM = c("group", "MEN1703_5uM", "DMSO"))

# colData(dds)$stim_2 <- factor(colData(dds)$stim_2)

DEG <- list()
for (i in 1:length(contrasts)) {
  DEG[[i]] <-
    as.data.frame(results(dds,
                          contrast = contrasts[[i]],
                          independentFiltering = TRUE, cooksCutoff = TRUE)) %>% tibble::rownames_to_column("SYMBOL") %>%
    dplyr::select(SYMBOL, log2FoldChange, pvalue, padj)
}

names(DEG) <- names(contrasts)

write.csv(DEG[[1]], file = "../data/tables/deg_exp_2_MEM_24h_3uM.csv")
write.csv(DEG[[2]], file = "../data/tables/deg_exp_2_MEM_24h_5uM.csv")

saveRDS(DEG, "../data/DEG_MEM_3_5_uM.RDS")

```

### Volcano plots

Volcano plots are provided for each contrast with coloring as indicated on the legend. Cutoffs for coloring are: FC = 1, padj = 0.05. Number of up/downregulated genes is indicated for significantly changed genes (padj < 0.05).

```{r volcano}

res <- bind_rows(DEG, .id = "sample")

write_csv(res, path = "../data/tables/deg_exp_2.csv")

temp1 <- nrow(dplyr::filter(res, sample == "DMSOvsMEN1703_3uM", log2FoldChange<0, padj<0.05))
temp2 <- nrow(dplyr::filter(res, sample == "DMSOvsMEN1703_3uM", log2FoldChange>0, padj<0.05))

temp3 <- nrow(dplyr::filter(res, sample == "DMSOvsMEN1703_5uM", log2FoldChange<0, padj<0.05))
temp4 <- nrow(dplyr::filter(res, sample == "DMSOvsMEN1703_5uM", log2FoldChange>0, padj<0.05))

dat_text <- data.frame(label = c(paste(temp1, "\ndownregulated"), 
                                 paste(temp2, "\nupregulated"), 
                                 paste(temp3, "\ndownregulated"), 
                                 paste(temp4, "\nupregulated")),
                    sample   = c("DMSOvsMEN1703_3uM", 
                                 "DMSOvsMEN1703_3uM", 
                                 "DMSOvsMEN1703_5uM", 
                                 "DMSOvsMEN1703_5uM"),
                    x = rep(c(-4, 4), 2),
                    y = rep(27, 4))

p1 <- EnhancedVolcano(res,
    lab = res$SYMBOL,
    selectLab = NA,
    x = 'log2FoldChange',
    y = 'padj',
    FCcutoff = 1,
    title = NULL,
    subtitle = NULL,
    pCutoff = 0.05,
    xlim = c(-7,7),
    ylim = c(0, 30),
    drawConnectors = TRUE,
    legendPosition = "bottom",
    legendLabSize = 10) +
    facet_wrap(~sample) + 
    theme(strip.background = element_rect(color="black", fill= "white"),
          text = element_text(size = 18)) + 
    geom_text(data = dat_text, mapping = aes(x = x, y = y, label = label))

p1
```


### Top differential genes heatmaps

To better visualize variable genes in each comparison, we are going to have a look at 25 most upregulated and 25 most downregulated genes in each contrasts and plot the valuse as a heatmap of sclaled (z-score) vsd values for each biological replicate.

```{r heatmaps_top, fig.asp=1}

topHeatmap <- function(x) {

top <- list()
for (i in 1:length(DEG)) {
top[[i]] <- rbind(top_n(DEG[[i]], x, log2FoldChange), top_n(DEG[[i]], -x, log2FoldChange))
}

vsd_table <- as.data.frame(assay(vsd))
vsd_scaled <<- t(scale(t(vsd_table)))
colnames(vsd_scaled) <- vsd$group

df1 <- vsd_scaled[rownames(vsd_scaled) %in% top[[1]]$SYMBOL, ]
df2 <- vsd_scaled[rownames(vsd_scaled) %in% top[[2]]$SYMBOL, ]

pheatmap(
  df1,
  show_rownames = TRUE,
  cluster_cols = FALSE,
  cellwidth = 10,
  cluster_rows = TRUE,
  scale = "none",
  angle_col = 270,
  labels_row = rownames(df1),
  main = paste0("top ", x, " up/down for DMSO vs MEN1703 3uM (vsd z-score)"))

pheatmap(
  df2,
  show_rownames = TRUE,
  cluster_cols = FALSE,
  cellwidth = 10,
  cluster_rows = TRUE,
  scale = "none",
  angle_col = 270,
  labels_row = rownames(df2),
  main = paste0("top ", x, " up/down for DMSO vs MEN1703 5uM (vsd z-score)"))

}

topHeatmap(25)


# genes <- c("CCL2", "CCL5", "CCL7", "CCL8", "CCL13", "CCL17",
#            "CCL18", "CCL24", "TGM2", "MMP1", "MMP7", "MMP9", 
#            "MMP12", "CD274", "VEGFA", "CHI3L2", "PDGFB", 
#            "TIMP1", "CHI3L1", "CHI3L2")

genes <- c("CCL2", "CCL13", "CCL17", "CCL24", "CCL28", "IL8", "IL15",
            "CHI3L1", "CHI3L2", "PDGFB", "VEGFA", "MMP9",
            "MMP25", "TGM2", "CD274", "CD209", "CD86", "MRC1",
            "IDO1", "IL4I1", "TIMP1", "PIM1", "PIM2.1", "PIM3")

df3 <- vsd_scaled[rownames(vsd_scaled) %in% genes, 1:6]

breaks <- c(seq(min(df3[,c(2:5)]), 0, by = 0.1), 
  seq(0, max(df3[,c(2:5)]), by = 0.1))

colors <- c(colorRampPalette(c("blue", "white"))(sum(breaks < 0)),
            colorRampPalette(c("white", "red"))(sum(breaks > 0)))

pheatmap(
  df3,
  show_rownames = TRUE,
  cluster_cols = FALSE,
  cellwidth = 10,
    cellheight = 10,
  cluster_rows = TRUE,
  scale = "none",
  angle_col = 270,
  labels_row = rownames(df3),
  color = colors,
  breaks = breaks)


deg_3 <- DEG[[1]]

rownames(deg_3) <- deg_3$SYMBOL
deg_3$SYMBOL <- NULL

deg_3_selected <- deg_3[rownames(deg_3) %in% genes, ]


breaks2 <- c(seq(min(deg_3_selected[, 1]), 0, by = 0.1), 
  seq(0, max(df3[,c(2:5)]), by = 0.1))

colors <- c(colorRampPalette(c("blue", "white"))(sum(breaks2 < 0)),
            colorRampPalette(c("white", "red"))(sum(breaks2 > 0)))

pheatmap(
  deg_3_selected[ , 1],
  show_rownames = TRUE,
  cluster_cols = FALSE,
  cellwidth = 10,
  cellheight = 10,
  cluster_rows = TRUE,
  scale = "none",
  angle_col = 270,
  labels_row = rownames(deg_3_selected),
  color = colors,
  breaks = breaks2)


```


```{r info}
sessionInfo()
```
